Allow creation of new (virtual) section from .pages

dimrozakis opened the pull request:
Allows the `nav` config of `.pages` to create new sections to organize
pages.

Fixes #37

dimrozakis commented:
Hi @lukasgeiter 

I created this PR because I wanted to reorganize the pages in a more nested hierarchy without changing their paths (and their links). This is working properly for me in the site I'm managing. All preexisting tests pass and I've added a couple of new tests for the new functionality.

Some notes:

1. I'm not very familiar with `mkdocs-awesome-pages-plugin`. This was a quick implementation which I'm sure could be improved upon. I'm not sure if you're happy with the use of nested functions. Or the term "virtual" section. Feel free to make suggestions or direct edits.
2. As previously said, this is working for me and all preexisting and new tests pass, but I haven't covered every possible use case in the tests. For example I don't test nested "virtual" sections. Or what happens if a virtual section has the same name as a real section. I can't 100% guarantee that this works properly in combination with every other feature of the plugin.
3. I haven't documented this feature in the README.
4. I'm not sure but I think that this can also satisfy the use case of https://github.com/lukasgeiter/mkdocs-awesome-pages-plugin#combine-custom-navigation--file-structure with a top level .pages instead of defining the nav section in mkdocs.yml.

Looking forward to your feedback. And thanks for creating this plugin.

lukasgeiter commented:
Hey @dimrozakis 

Thanks for the PR! Right now I'm a bit short on time, but I'll take a closer look in the coming days.

lukasgeiter commented on mkdocs_awesome_pages_plugin/meta.py:
```diff
@@ -34,8 +34,11 @@ def from_yaml(item: Union[str, dict], context: str):
+                if isinstance(value, str):
+                    return MetaNavItem(value, title)
+                elif isinstance(value, list):
+                    return MetaNavItem([MetaNavItem.from_yaml(it, context) for it in value], title)
```
The type annotation of `value` needs to be extended for this. `Union[str, List[MetaNavItem]]` should do the trick.

lukasgeiter commented on mkdocs_awesome_pages_plugin/navigation.py:
```diff
@@ -37,6 +37,10 @@ def __init__(self, filename: str):
         )
 
 
+class VSection(Section):
```
Could you spell this out as `VirtualSection`? Please also do so for any other instances of the term.

lukasgeiter commented on mkdocs_awesome_pages_plugin/navigation.py:
```diff
@@ -86,19 +90,25 @@ def _nav(self, items: List[NavigationItem], meta: Meta) -> List[NavigationItem]:
+                    rest_items.append(meta_item)
+                    result.append(meta_item)
+
+                elif isinstance(meta_item.value, (list, tuple)):
```
Is `tuple` needed here?

lukasgeiter commented on mkdocs_awesome_pages_plugin/navigation.py:
```diff
@@ -86,19 +90,25 @@ def _nav(self, items: List[NavigationItem], meta: Meta) -> List[NavigationItem]:
-                result.append(meta_item)
-            else:
-                if meta_item.value in items_by_basename:
+        def _make_nav_rec(meta_nav):
```
Please add a type annotation to the parameter here

lukasgeiter commented on mkdocs_awesome_pages_plugin/navigation.py:
```diff
@@ -109,21 +119,30 @@ def _nav(self, items: List[NavigationItem], meta: Meta) -> List[NavigationItem]:
-            for index, item in enumerate(result):
-                if isinstance(item, MetaNavRestItem):
-                    result[index : index + 1] = rest[item]
+            def _expand_rest_rec(result):
```
Please add a type annotation to the parameter here

lukasgeiter reviewed (changes requested):
Hey there, sorry for the long wait! There are few stylistic changes I'd like you to make, but otherwise it's looking pretty good 👍

dimrozakis commented:
@lukasgeiter thanks for the review. I've addressed all your comments. Any thoughts on points 2,3,4 from https://github.com/lukasgeiter/mkdocs-awesome-pages-plugin/pull/59#issuecomment-1021559894?

lukasgeiter reviewed (approved):
Great, thanks for the changes! My thoughts on your previous comment:

2. I think it's fine. I do appreciate your concern though!
3. No worries, I'll take care of that.
4. Almost, but not quite. There's still a difference in which files can be targeted. A `.pages` can only access the files and folders in the same directory, while  `mkdocs.yml` has access to everything.
